types/mempool/mempool_test.go

```diff
@@ -0,0 +1,659 @@
+		return nil, fmt.Errorf("error")
+	}}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+	tx.getSigs = func() ([]txsigning.SignatureV2, error) {
+		return nil, nil
+	}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+
+	// removing a tx not in the mempool should error
+	mp = mempool.NewPriorityMempool()
+	require.NoError(t, mp.Insert(ctx, txs[0]))
+	require.ErrorIs(t, mp.Remove(txs[1]), mempool.ErrTxNotFound)
+	mempool.DebugPrintKeys(mp)
+
+	// inserting a tx with a different priority should overwrite the old tx
+	newPriorityTx := testTx{
+		address:  txs[0].address,
+		priority: txs[0].priority + 1,
+		nonce:    txs[0].nonce,
+	}
+	require.NoError(t, mp.Insert(ctx, newPriorityTx))
+	require.Equal(t, 1, mp.CountTx())
+}
+
+type txSpec struct {
+	i int
+	h int
+	p int
+	n int
+	a sdk.AccAddress
+}
+
+func (tx txSpec) String() string {
+	return fmt.Sprintf("[tx i: %d, a: %s, p: %d, n: %d]", tx.i, tx.a, tx.p, tx.n)
+}
+
+func TestOutOfOrder(t *testing.T) {
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 2)
+	sa := accounts[0].Address
+	sb := accounts[1].Address
+
+	outOfOrders := [][]testTx{
+		{
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		},
+		{
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		}}
+
+	for _, outOfOrder := range outOfOrders {
+		var mtxs []mempool.Tx
+		for _, mtx := range outOfOrder {
+			mtxs = append(mtxs, mtx)
+		}
+		err := validateOrder(mtxs)
+		require.Error(t, err)
+	}
+
+	seed := time.Now().UnixNano()
+	randomTxs := genRandomTxs(seed, 1000, 10)
+	var rmtxs []mempool.Tx
+	for _, rtx := range randomTxs {
+		rmtxs = append(rmtxs, rtx)
+	}
+
+	require.Error(t, validateOrder(rmtxs))
+
+}
+
+func (s *MempoolTestSuite) TestTxOrder() {
+	t := s.T()
+	ctx := sdk.NewContext(nil, tmproto.Header{}, false, log.NewNopLogger())
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 5)
+	sa := accounts[0].Address
+	sb := accounts[1].Address
+	sc := accounts[2].Address
+
+	tests := []struct {
+		txs   []txSpec
+		order []int
+		fail  bool
+	}{
+		{
+			txs: []txSpec{
+				{p: 21, n: 4, a: sa},
+				{p: 8, n: 3, a: sa},
+				{p: 6, n: 2, a: sa},
+				{p: 15, n: 1, a: sb},
+				{p: 20, n: 1, a: sa},
+			},
+			order: []int{4, 3, 2, 1, 0},
+		},
+		{
+			txs: []txSpec{
+				{p: 3, n: 0, a: sa},
+				{p: 5, n: 1, a: sa},
+				{p: 9, n: 2, a: sa},
+				{p: 6, n: 0, a: sb},
+				{p: 5, n: 1, a: sb},
+				{p: 8, n: 2, a: sb},
+			},
+			order: []int{3, 4, 5, 0, 1, 2},
+		},
+		{
+			txs: []txSpec{
+				{p: 21, n: 4, a: sa},
+				{p: 15, n: 1, a: sb},
+				{p: 20, n: 1, a: sa},
+			},
+			order: []int{2, 0, 1}},
```

Comment:
nit:

```suggestion
			order: []int{2, 0, 1}
		},
```

Reply:
Deleting this comma I see
```
mempool_test.go:263:26: missing ',' before newline in composite literal
```

Reply:
The change here was actually to move the second `}` to a new line, to align with other test cases.